wallet/wallet_v4.tact

```diff
@@ -0,0 +1,121 @@
+interface Plugin {
+    internal requested_funds#f06c7567(query_id: Uint64);
+    internal notify_removed#e4737472(query_id: Uint64);
+    internal notify_installed#6e6f7465(query_id: Uint64);
+    internal notify_removed#64737472(query_id: Uint64);
+}
+
+contract Wallet {
+    var seqno: Uint32
+    var subwallet: Uint32
+    var pubkey: Pubkey
+    var plugins: HashSet[Plugin]
+
+    bounced default() {
+        // ignore bounced messages
+    }
+
+    internal default() {
+        // ignore simple transfers
+    }
+
+    internal request_funds#706c7567(
+        query_id: Uint64,
+        requested_amount: Coins,
+        requested_extra: ExtraCurrencies,
+        msg.sender: Plugin,
+    ) {
+        verify(self.balance() - msg.amount >= requested_amount, "Not enough funds");
+
+        send(InboundCoins)
+        msg.sender.requested_funds(
+            query_id,
+            msg.amount: requested_amount,
+            msg.extra: requested_extra,
+        );
+    }
+
+    internal remove_plugin#64737472(query_id: Uint64, msg.sender: Plugin) {
+        self.remove_plugin_from_list(msg.sender);
+
+        if msg.bounceable {
+            send(InboundCoins) msg.sender.notify_removed(query_id);
+        }
+    }
+
+    external pre_condition(body: Slice) -> Slice {
+        let sign = body.read(Signature);
+        if sign.verify(body, self.pubkey) {
+            self.accept();
+            self.seqno += 1;
+            self.commit();
+            return body;
+        } else {
+            thrown("Invalid signature");
+        }
+    }
+
+    external simple_transfer#0(messages: List[(Cell, Sendmode)]) -> Slice {
```

Comment:
This is a tricky one. We want to tie the list enumeration to the refs. 

Option A: if we don't have a clean type language to describe this, then i'd keep a raw loop that reads refs and also scans for flags.

```
external simple_transfer#00() {
    for raw_msg in msg.body.refs() {
        let flags = msg.body.read_byte();
        send_raw(raw_msg, flags);
    }
}

```

PS. We should treat `#0` as a 4-bit zero and `#00` as a 8-bit zero. TL-B examples show that `#000` is a 12-bit tag.